### PR TITLE
arc: interrupts: Explain return from interrupt to cooperative thread

### PR DESCRIPTION
**Pull request by abrodkin**
The code in question is very non-trivial so without good explanation
it takes a lot of time to realize what's done there and why
it still works in the end.

Here I'm trying to save a couple of man-days for the next developers
who's going to touch that piece of code.

### CODE REVIEW

File: arch/arc/core/regular_irq.S [hunk 1]
```diff
@@ -35,6 +35,168 @@ GTEXT(_rirq_common_interrupt_swap)
  * TODO: Revist this if FIRQ becomes configurable.
```
Comment (ruuddw) [inline arch/arc/core/regular_irq.S]: I think this TODO doesn't hold anymore, FIRQ is configurable now. Does it mean the NUM_REGULAR_IRQ_PRIO_LEVELS isn't corrent in that case?

File: arch/arc/core/regular_irq.S [hunk 2]
```diff
@@ -35,6 +35,168 @@ GTEXT(_rirq_common_interrupt_swap)
...
+/*
+
+===========================================================
+ RETURN FROM INTERRUPT TO COOPERATIVE THREAD
+===========================================================
+
+That's a special case because:
+ 1. We return from IRQ handler to a cooperative thread
+ 2. During IRQ handling context switch did happen
+ 3. Thread we're returining to previously voluntarily gave control
```
Comment (ruuddw) [inline arch/arc/core/regular_irq.S]: some typo's, suggest to use : ... "returning to a thread which previously gave control"

File: arch/arc/core/regular_irq.S [hunk 3]
```diff
@@ -35,6 +35,168 @@ GTEXT(_rirq_common_interrupt_swap)
...
+    to another thread because of:
+     - Falling k_sleep()
```
Comment (ruuddw) [inline arch/arc/core/regular_irq.S]: Calling, not Falling, I assume

File: arch/arc/core/regular_irq.S [hunk 4]
```diff
@@ -35,6 +35,168 @@ GTEXT(_rirq_common_interrupt_swap)
...
+     - Bumping into locked sync primitive etc
+
+What (3) means is before passing control to another thread our thread
+in question:
+ a. Stashed all precious caller-saved registers on its stack
+ b. Pushed return address to the top of the stack as well
+
+That's how thread's stack looks like right before jumping to another thread:
+----------------------------->8---------------------------------
+PRE-COTEXT-SWITCH STACK
```
Comment (ruuddw) [inline arch/arc/core/regular_irq.S]: CONTEXT

File: arch/arc/core/regular_irq.S (hunk 3 above)
Comment (abrodkin) [reply arch/arc/core/regular_irq.S]: I meant exactly "falling" in a context "falling asleep", this one even sounds similarly "falling ka-sleep" :)

Comment (ruuddw) [reply arch/arc/core/regular_irq.S]: yeah, I though about that. But, falling could also be mistaken/misread for failing...

File: arch/arc/core/regular_irq.S (hunk 1 above)
Comment (abrodkin) [reply arch/arc/core/regular_irq.S]: I guess this is orthogonal to this particular pull-request and thus should be addressed separately as I'm not touching any topics outside complex return from IRQ.

File: arch/arc/core/regular_irq.S (hunk 2 above)
Comment (abrodkin) [reply arch/arc/core/regular_irq.S]: Ok, good. Will change that.

**Comment by abrodkin** [comment]
Pushed a re-spin with changes proposed by @ruuddw.

**Review by ruuddw — APPROVED** [review]
Thanks, these explanations are really useful, certainly for people not so familiar with ARC and Zephyr context switching details.

**Review by vonhust — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**